content/blog/c-off-ramp.md

```diff
@@ -0,0 +1,218 @@
+    return newpath;
+}
+```
+
+This is tricky. First, it calls the “real” `relocate` (but using its Vala name), and stores the result in an `unowned` string. So far, `valac` will not allocate any memory.
+
+We then test the value returned, and if it is unequal to the original argument, return it. Because we use the `return_if_fail` macro, `valac` will not insert memory allocation code at this point, but that’s OK, because `relocate` has already `malloc`ed in this case, so the return value is an allocated `string` as indicated.
+
+Otherwise, we simply `return` the un-`malloc`ed value. In the generated C, `valac` will insert a `g_strdup` at this point, so again the return value is an allocated string.
+
+Phew!
+
+##### `strfreev`
+
+Rather surprisingly, while `g_strfreev` takes `char **` (because it frees a list of strings), `strfreev` takes `string **`. Why? Well, Vala uses one level of pointer indirection to indicate that a value is not managed by Vala. So we have the following set of equivalences:
+
+* `string` = `char *` (generates `char *`; as we expect, a string is a pointer to `char`)
+* `string *` = `char *` (generates `char *`; this is just a `string` that Vala does not manage)
+* `string **` = `char **` (generates `char **`; now we have a list of `string`)
+
+This was a problem for me because I wanted to use `strfreev` to free a “real” `char **` passed in from C. But `valac` gives a type error. I got sufficiently confused that I submitted a [merge request](https://gitlab.gnome.org/GNOME/vala/-/merge_requests/383), and Rico had to de-confuse me. In the end, all I needed was a cast:
+
+```vala
+strfreev((string **)c_string_list);
+```
+
+The cast placates `valac`, and the generated C ends up calling `g_strfreev`, which expected a `char **` all along.
+
+##### `[s]size_t` is not really `[s]size_t`
+
+When compiling my Vala version of Enchant on Windows, I accidentally tried compiling it in a flavour of Windows I don’t normally test, namely Mingw MSYS (I normally build for Mingw64-x86_64 and Mingw64-i686). To my surprise, I got a compilation error: GCC complained that one 8-byte signed integer type (`long int`) was incompatible with another 8-byte signed integer type (`long long int`). Yes, these are both 64-bit types on this system, and yes, they are nonetheless incompatible.
+
+But why did I have this problem in the first place? I was using Vala’s `ssize_t`, and Vala maps this to GLib’s `gsize`. I don’t know why, but it’s a problem, because in some cases, like mine, these two types end up with different fundamental types in the C compiler, and an error results. I filed an [issue](https://gitlab.gnome.org/GNOME/vala/-/issues/1539) and found a workaround, which was to copy the definition of `size_t` and `ssize_t` from `posix.vapi`, which are normally only used when `valac` is used with `--profile posix` and the GLib definitions are not used. The `posix.vapi` definitions correctly map `size_t` in Vala to `size_t` in C.
+
```

Comment:
maybe mention that this issue is only due to the preexisting API. (afai understood)

Reply:
I don't understand this comment. As I've said elsewhere, mapping `size_t` to another type (that is not itself a synonym for `size_t`, at least) is simply wrong. This would be a problem with any API involving `size_t` (i.e. a non-GLib API) on any platform that has `gsize` different from `size_t`. I understand that this is an unusual situation and I could easily imagine myself having been the person who thought that this was OK (except why use the name `size_t` and not, say, `usize` or some other more Vala-ish name—`size` would be logical, by the usual Vala rule of removing a leading `g`?), but it is, simply, wrong. Arguably it's a bug in GLib, not Vala (i.e. `gsize` *should* always be the same as `size_t`). But it's definitely a bug!

Reply:
I see, then maybe point very briefly out that the issue is actually in glib, not vala.

Reply:
I think there really is a problem in Vala: Vala should not use the name `size_t` to mean something other than `size_t`. That's the problem I describe in the post. The issue of whether GLib, having invented `gsize`, should tie it to `size_t` or not is a separable issue. If Vala had used the name `size` for `gsize`, then it would be a GLib problem, and in Vala one could use `size_t` to mean `size_t`, but this is currently not an option.